lib/src/usage_impl.dart

```diff
@@ -219,36 +254,66 @@ class AnalyticsImpl implements Analytics {
   ///
   /// Valid values for [hitType] are: 'pageview', 'screenview', 'event',
   /// 'transaction', 'item', 'social', 'exception', and 'timing'.
-  Future sendRaw(String hitType, Map<String, dynamic> args) {
-    return _sendPayload(hitType, args);
+  Future sendRaw(String hitType, Map<String, String> args) {
+    return _enqueuePayload(hitType, args);
   }
 
+  /// Puts a single event in the queue. If batching is not enabled it will be
+  /// send immediately - otherwise when the batch is full (20 events) or when
+  /// the batching callback is over.
   /// Valid values for [hitType] are: 'pageview', 'screenview', 'event',
   /// 'transaction', 'item', 'social', 'exception', and 'timing'.
-  Future _sendPayload(String hitType, Map<String, dynamic> args) {
-    if (!enabled) return Future.value();
+  Future<void> _enqueuePayload(String hitType, Map<String, String> args) async {
+    if (!enabled) return;
+
+    final eventArgs = <String, String>{
+      ...args,
+      ..._variableMap,
+      'v': '1', // protocol version
+      'tid': trackingId,
+      'cid': clientId,
+      't': hitType,
+    };
+
+    final batch = <Map<String, String>>[];
+
+    // See if we currently are batching events:
+    final batchingQueue = Zone.current[_batchingKey];
+    if (batchingQueue is _BatchingQueue) {
+      // Add the current event to the batch.
+      batchingQueue.enqueuedEvents.add(eventArgs);
+      if (!batchingQueue.isFull) {
+        // Queue not full yet. Do nothing.
+        return;
+      }
+      // We have a full batch, start a new one.
+      batch.addAll(batchingQueue.takeBatch());
+    } else {
+      batch.add(eventArgs);
+    }
+    _trySendBatch(batch);
+  }
 
+  void _trySendBatch(List<Map<String, String>> batch) {
     if (_bucket.removeDrop()) {
-      _variableMap.forEach((key, value) {
-        args[key] = value;
-      });
-
-      args['v'] = '1'; // protocol version
-      args['tid'] = trackingId;
-      args['cid'] = clientId;
-      args['t'] = hitType;
-
-      _sendController.add(args);
+      for (final args in batch) {
+        _sendController.add(args);
+      }
 
-      return _recordFuture(postHandler.sendPost(_url, args));
-    } else {
-      return Future.value();
+      // See if we currently are batching events:
+      final batchingQueue = Zone.current[_batchingKey];
+      if (batchingQueue is _BatchingQueue) {
```

Comment:
why not use `batch.length == 1` to decide which end-point to use?

Reply:
Yeah - not sure - I thought it would be more consistent that everything goes to the batching endpoint within a `withBatching` but I'm not sure it matters much.

Reply:
Now it will send to `/collect` if there is only one message and `/batch` otherwise.